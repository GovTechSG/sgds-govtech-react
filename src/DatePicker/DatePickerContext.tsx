import * as React from 'react';

export type CalendarView = 'day' | 'month' | 'year';
export interface DatePickerContextProps {
  view: CalendarView;
  setView: React.Dispatch<React.SetStateAction<CalendarView>>;
  focusedDateIndex: number;
  setFocusedDateIndex: React.Dispatch<React.SetStateAction<number>>;
  focusedMonthIndex: number;
  setFocusedMonthIndex: React.Dispatch<React.SetStateAction<number>>;
  focusedYearIndex: number;
  setFocusedYearIndex: React.Dispatch<React.SetStateAction<number>>;
  yearPositionIndex: number; // to keep track of the position of the selected year, without this, year range position will not be consistent after selection
  setYearPositionIndex: React.Dispatch<React.SetStateAction<number>>;
}
const context = React.createContext<DatePickerContextProps>({
  view: 'day',
  setView: () => {},
  focusedDateIndex: 0,
  setFocusedDateIndex: () => {},
  focusedMonthIndex: 0,
  setFocusedMonthIndex: () => {},
  focusedYearIndex: 0,
  setFocusedYearIndex: () => {},
  yearPositionIndex: 0,
  setYearPositionIndex: () => {},
});

context.displayName = 'DatePickerContext';

export default context;
