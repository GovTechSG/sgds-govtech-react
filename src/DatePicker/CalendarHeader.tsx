import * as React from 'react';
import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
interface CalendarHeaderProps {
  displayDate: Date;
  onChange: (date: Date) => void;
  monthLabels: string[];
}
const MONTH_LABELS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

const CalendarHeader: React.FC<CalendarHeaderProps> = ({
  monthLabels = MONTH_LABELS,
  ...props
}) => {
  const { view, setView } = useContext(DatePickerContext);

  const handleClickPrevious = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);
    if (view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 1);
      props.onChange(newDisplayDate);
    }

    if (view === 'year') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 10);
      props.onChange(newDisplayDate);
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() - 1);
    props.onChange(newDisplayDate);
  };

  const handleClickNext = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);

    if (view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 1);
      props.onChange(newDisplayDate);
    }

    if (view === 'year') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 10);
      props.onChange(newDisplayDate);
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() + 1);
    props.onChange(newDisplayDate);
  };

  const changeView = () => {
    switch (view) {
      case 'day':
        setView('month');
        break;
      case 'month':
        setView('year');
        break;
      case 'year':
        break;
    }
  };
  const renderHeader = () => {
    if (view === 'month') return `${props.displayDate.getFullYear()}`;
    if (view === 'year')
      return `${props.displayDate.getFullYear() - 5} - ${
        props.displayDate.getFullYear() + 6
      }`;
    return `${
      monthLabels[props.displayDate.getMonth()]
    } ${props.displayDate.getFullYear()}`;
  };
  return (
    <div className="text-center d-flex justify-content-around">
      <div
        className="text-muted pull-left"
        onClick={handleClickPrevious}
        style={{ cursor: 'pointer' }}
      >
        <i className="bi bi-chevron-left"></i>
      </div>
      <button onClick={changeView} disabled={view === 'year'}>
        {renderHeader()}
      </button>
      <div
        className="text-muted pull-right"
        onClick={handleClickNext}
        style={{ cursor: 'pointer' }}
      >
        <i className="bi bi-chevron-right"></i>
      </div>
    </div>
  );
};
export default CalendarHeader;
