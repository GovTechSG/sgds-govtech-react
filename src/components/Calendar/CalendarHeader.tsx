import Button from '@restart/ui/esm/Button';
import * as React from 'react';
import Dropdown from '../Dropdown/Dropdown';
import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
interface CalendarHeaderProps {
  displayDate: Date;
  minDate?: string;
  maxDate?: string;
  onChange: (date: Date) => void;
  monthLabels: string[];
  previousButtonElement: string | JSX.Element;
  nextButtonElement: string | JSX.Element;
  years?: number[];
}
const MONTH_LABELS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];
const YEARS = (): number[] => {
  const currentYear = new Date().getFullYear();
  const yearsArray = [];
  const start = currentYear - 50;
  const end = currentYear + 50;
  for (let i = start; i < end; i++) {
    yearsArray.push(i);
  }
  return yearsArray;
};
const CalendarHeader: React.FC<CalendarHeaderProps> = ({
  monthLabels = MONTH_LABELS,
  previousButtonElement = '<',
  nextButtonElement = '>',
  years = YEARS(),
  ...props
}) => {
  const { view, setView } = useContext(DatePickerContext);
  const [year, setYear] = React.useState(props.displayDate.getFullYear());
  const displayingMinMonth = () => {
    if (!props.minDate) return false;

    const displayDate = new Date(props.displayDate);
    const minimumDate = new Date(props.minDate);
    return (
      minimumDate.getFullYear() == displayDate.getFullYear() &&
      minimumDate.getMonth() == displayDate.getMonth()
    );
  };

  const displayingMaxMonth = () => {
    if (!props.maxDate) return false;
    const displayDate = new Date(props.displayDate);
    const maximumDate = new Date(props.maxDate);
    return (
      maximumDate.getFullYear() == displayDate.getFullYear() &&
      maximumDate.getMonth() == displayDate.getMonth()
    );
  };

  const handleClickPrevious = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);
    if(view === 'month'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 1)
      props.onChange(newDisplayDate)
    }

    if(view === 'year'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() - 10)
      props.onChange(newDisplayDate)
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() - 1);
    props.onChange(newDisplayDate);
  };

  const handleClickNext = () => {
    const newDisplayDate = new Date(props.displayDate);
    newDisplayDate.setDate(1);

    if(view === 'month') {
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 1)
      props.onChange(newDisplayDate)
    }

    if(view === 'year'){
      newDisplayDate.setFullYear(newDisplayDate.getFullYear() + 10)
      props.onChange(newDisplayDate)
    }

    newDisplayDate.setMonth(newDisplayDate.getMonth() + 1);
    props.onChange(newDisplayDate);
  };
  const handleSelectYear = (key: string | null) => {
    const newDate = props.displayDate;
    newDate.setFullYear(parseInt(key!));
    props.onChange(newDate);
    setYear(parseInt(key!));
  };
  const [header, setHeader] = React.useState(
    `${monthLabels[props.displayDate.getMonth()]} ${year}`
  );

  const changeView = () => {
    switch (view) {
      case 'day':
        setView('month');
        setHeader(`${year}`);
        break;
      case 'month':
        setView('year');
        break;
      case 'year':
        break;
    }
  };
  const renderHeader = () => {
    if (view === 'month') return `${props.displayDate.getFullYear()}`;
    if (view === 'year') return `${props.displayDate.getFullYear() - 5} - ${props.displayDate.getFullYear() + 6}`;
    return `${monthLabels[props.displayDate.getMonth()]} ${props.displayDate.getFullYear()}`;
  };
  return (
    <div className="text-center d-flex justify-content-around">
      <div
        className="text-muted pull-left"
        onClick={handleClickPrevious}
        style={{ cursor: 'pointer' }}
      >
        {displayingMinMonth() ? null : previousButtonElement}
      </div>
      <Button onClick={changeView} disabled={view === 'year'}>
        {renderHeader()}
      </Button>
      {/*    <Dropdown onSelect={handleSelectYear}>
        <Dropdown.Toggle variant="white" >
          {year}
        </Dropdown.Toggle>
        <Dropdown.Menu >
          {years.map((yr: number) => (
            <Dropdown.Item
            key={yr}
            active={yr===year}
              eventKey={yr}
            >
              {yr}
            </Dropdown.Item>
          ))}
        </Dropdown.Menu>
      </Dropdown> */}
      <div
        className="text-muted pull-right"
        onClick={handleClickNext}
        style={{ cursor: 'pointer' }}
      >
        {displayingMaxMonth() ? null : nextButtonElement}
      </div>
    </div>
  );
};
export default CalendarHeader;
