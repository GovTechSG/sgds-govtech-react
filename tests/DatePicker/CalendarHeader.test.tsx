import * as React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import CalendarHeader, {
  MONTH_LABELS,
} from '../../src/DatePicker/CalendarHeader';
import DatePickerContext, {
  CalendarView,
} from '../../src/DatePicker/DatePickerContext';

describe('CalendarHeader', () => {
  const mockFn = jest.fn();
  const resetFocusOnHeader = jest.fn();
  const handleTabPressOnPreviousButton = jest.fn();
  const handleTabPressOnHeaderTitle = jest.fn();
  const handleTabPressOnNextButton = jest.fn();
  const displayDate = new Date();
  const displayMonth = MONTH_LABELS[displayDate.getMonth()];
  const displayYear = displayDate.getFullYear();
  const contextValue = {
    view: 'day' as CalendarView,
    setView: jest.fn(),
    focusedDateIndex: 0,
    setFocusedDateIndex: jest.fn(),
    focusedMonthIndex: 0,
    setFocusedMonthIndex: jest.fn(),
    focusedYearIndex: 0,
    setFocusedYearIndex: jest.fn(),
    yearPositionIndex: 0,
    setYearPositionIndex: jest.fn(),
  };

  it('should have default html structure', () => {
    const displayDate = new Date(2022, 2, 20);
    const displayMonth = MONTH_LABELS[displayDate.getMonth()];
    const displayYear = displayDate.getFullYear();
    const { asFragment, container, getByText } = render(
      <CalendarHeader
        onChange={mockFn}
        displayDate={displayDate}
        resetFocusOnHeader={resetFocusOnHeader}
        handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
        handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
        handleTabPressOnNextButton={handleTabPressOnNextButton}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(
      container.querySelector('div.d-flex.text-center.justify-content-between')
    ).toBeInTheDocument();
    expect(
      container.querySelector('div>button>i.bi.bi-chevron-left')
    ).toBeInTheDocument();
    expect(container.querySelectorAll('button').length).toEqual(3);
    expect(
      container.querySelector('div>button>i.bi.bi-chevron-right')
    ).toBeInTheDocument();

    expect(getByText(`${displayMonth} ${displayYear}`)).toBeInTheDocument();
  });

  it('click calendarHeader should trigger setView call;', async () => {
    const { container } = render(
      <DatePickerContext.Provider value={contextValue}>
        <CalendarHeader
          onChange={mockFn}
          displayDate={displayDate}
          resetFocusOnHeader={resetFocusOnHeader}
          handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
          handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
          handleTabPressOnNextButton={handleTabPressOnNextButton}
        />
      </DatePickerContext.Provider>
    );
    const headerButton = container.querySelectorAll('button')[1];
    expect(headerButton?.textContent).toEqual(`${displayMonth} ${displayYear}`);

    fireEvent.click(headerButton!);

    await waitFor(() => {
      expect(contextValue.setView).toHaveBeenCalled();
    });
  });
  it('given context view month, calendar header shows full year;', async () => {
    const contextValue = {
      view: 'month' as CalendarView,
      setView: jest.fn(),
      focusedDateIndex: 0,
      setFocusedDateIndex: jest.fn(),
      focusedMonthIndex: 0,
      setFocusedMonthIndex: jest.fn(),
      focusedYearIndex: 0,
      setFocusedYearIndex: jest.fn(),
      yearPositionIndex: 0,
      setYearPositionIndex: jest.fn(),
    };

    const { container } = render(
      <DatePickerContext.Provider value={contextValue}>
        <CalendarHeader
          onChange={mockFn}
          displayDate={displayDate}
          resetFocusOnHeader={resetFocusOnHeader}
          handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
          handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
          handleTabPressOnNextButton={handleTabPressOnNextButton}
        />
      </DatePickerContext.Provider>
    );
    const headerButton = container.querySelectorAll('button')[1];
    expect(headerButton?.textContent).not.toEqual(
      `${displayMonth} ${displayYear}`
    );
    expect(headerButton?.textContent).toEqual(`${displayYear}`);

    fireEvent.click(headerButton!);

    await waitFor(() => {
      expect(contextValue.setView).toHaveBeenCalled();
    });
  });
  it('given context view year, calendar header shows full year;', async () => {
    const contextValue = {
      view: 'year' as CalendarView,
      setView: jest.fn(),
      focusedDateIndex: 0,
      setFocusedDateIndex: jest.fn(),
      focusedMonthIndex: 0,
      setFocusedMonthIndex: jest.fn(),
      focusedYearIndex: 0,
      setFocusedYearIndex: jest.fn(),
      yearPositionIndex: 0,
      setYearPositionIndex: jest.fn(),
    };

    const { container } = render(
      <DatePickerContext.Provider value={contextValue}>
        <CalendarHeader
          onChange={mockFn}
          displayDate={displayDate}
          resetFocusOnHeader={resetFocusOnHeader}
          handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
          handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
          handleTabPressOnNextButton={handleTabPressOnNextButton}
        />
      </DatePickerContext.Provider>
    );
    const headerButton = container.querySelectorAll('button')[1];
    expect(headerButton?.textContent).not.toEqual(
      `${displayMonth} ${displayYear}`
    );
    expect(headerButton?.textContent).not.toEqual(`${displayYear}`);
    expect(headerButton?.textContent).toEqual(
      `${displayDate.getFullYear()} - ${displayDate.getFullYear() + 11}`
    );

    fireEvent.click(headerButton!);

    await waitFor(() => {
      expect(contextValue.setView).not.toHaveBeenCalled();
    });
  });

  it('onclick bi-chevron-left should trigger mockFn ', async () => {
    const { container } = render(
      <CalendarHeader
        onChange={mockFn}
        displayDate={displayDate}
        resetFocusOnHeader={resetFocusOnHeader}
        handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
        handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
        handleTabPressOnNextButton={handleTabPressOnNextButton}
      />
    );
    const previousIcon = container.querySelector('i.bi-chevron-left');

    expect(previousIcon).toBeInTheDocument();
    fireEvent.click(previousIcon!);

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalled();
    });
  });
  it('onclick bi-chevron-right should trigger mockFn', async () => {
    const { container } = render(
      <CalendarHeader
        onChange={mockFn}
        displayDate={displayDate}
        resetFocusOnHeader={resetFocusOnHeader}
        handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
        handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
        handleTabPressOnNextButton={handleTabPressOnNextButton}
      />
    );
    const nextIcon = container.querySelector('i.bi-chevron-right');

    expect(nextIcon).toBeInTheDocument();
    fireEvent.click(nextIcon!);

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalled();
    });
  });
});

describe('CalendarHeaderComponent day view in usage', () => {
  const CalendarHeaderComponent = () => {
    const [date, setDate] = React.useState(new Date(2022, 2, 20));
    const onChangeMonth = (newDisplayDate: Date) => {
      setDate(newDisplayDate);
    };
    const resetFocusOnHeader = jest.fn();
    const handleTabPressOnPreviousButton = jest.fn();
    const handleTabPressOnHeaderTitle = jest.fn();
    const handleTabPressOnNextButton = jest.fn();

    return (
      <CalendarHeader
        displayDate={date}
        onChange={onChangeMonth}
        resetFocusOnHeader={resetFocusOnHeader}
        handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
        handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
        handleTabPressOnNextButton={handleTabPressOnNextButton}
      />
    );
  };

  it('should change month when click previous', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-left')).toBeInTheDocument();
    expect(getByText('March 2022')).toBeInTheDocument();

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);

    await waitFor(() => expect(getByText('February 2022')).toBeInTheDocument());
    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    await waitFor(() => expect(getByText('January 2022')).toBeInTheDocument());

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    await waitFor(() => expect(getByText('December 2021')).toBeInTheDocument());
    fireEvent.click(container.querySelector('i.bi-chevron-right')!);
    await waitFor(() => expect(getByText('January 2022')).toBeInTheDocument());
  });
  it('should change month when click next', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-right')).toBeInTheDocument();
    expect(getByText('March 2022')).toBeInTheDocument();

    fireEvent.click(container.querySelector('i.bi-chevron-right')!);

    await waitFor(() => expect(getByText('April 2022')).toBeInTheDocument());
  });
});
describe('CalendarHeaderComponent month view in usage', () => {
  const CalendarHeaderComponent = () => {
    const [date, setDate] = React.useState(new Date(2022, 2, 20));
    const onChangeMonth = (newDisplayDate: Date) => {
      setDate(newDisplayDate);
    };
    const resetFocusOnHeader = jest.fn();
    const handleTabPressOnPreviousButton = jest.fn();
    const handleTabPressOnHeaderTitle = jest.fn();
    const handleTabPressOnNextButton = jest.fn();
    const contextValue = {
      view: 'month' as CalendarView,
      setView: jest.fn(),
      focusedDateIndex: 0,
      setFocusedDateIndex: jest.fn(),
      focusedMonthIndex: 0,
      setFocusedMonthIndex: jest.fn(),
      focusedYearIndex: 0,
      setFocusedYearIndex: jest.fn(),
      yearPositionIndex: 0,
      setYearPositionIndex: jest.fn(),
    };
    return (
      <DatePickerContext.Provider value={contextValue}>
        <CalendarHeader
          displayDate={date}
          onChange={onChangeMonth}
          resetFocusOnHeader={resetFocusOnHeader}
          handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
          handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
          handleTabPressOnNextButton={handleTabPressOnNextButton}
        />
        ;
      </DatePickerContext.Provider>
    );
  };

  it('should change month when click previous', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-left')).toBeInTheDocument();
    expect(getByText('2022')).toBeInTheDocument();

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);

    await waitFor(() => expect(getByText('2021')).toBeInTheDocument());
    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    await waitFor(() => expect(getByText('2020')).toBeInTheDocument());

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    await waitFor(() => expect(getByText('2019')).toBeInTheDocument());
    fireEvent.click(container.querySelector('i.bi-chevron-right')!);
    await waitFor(() => expect(getByText('2020')).toBeInTheDocument());
  });
  it('should change month when click next', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-right')).toBeInTheDocument();
    expect(getByText('2022')).toBeInTheDocument();

    fireEvent.click(container.querySelector('i.bi-chevron-right')!);

    await waitFor(() => expect(getByText('2023')).toBeInTheDocument());
  });
});
describe('CalendarHeaderComponent year view in usage', () => {
  const displayDate = new Date();
  const CalendarHeaderComponent = () => {
    const [date, setDate] = React.useState(displayDate);
    const onChangeMonth = (newDisplayDate: Date) => {
      setDate(newDisplayDate);
    };
    const resetFocusOnHeader = jest.fn();
    const handleTabPressOnPreviousButton = jest.fn();
    const handleTabPressOnHeaderTitle = jest.fn();
    const handleTabPressOnNextButton = jest.fn();
    const contextValue = {
      view: 'year' as CalendarView,
      setView: jest.fn(),
      focusedDateIndex: 0,
      setFocusedDateIndex: jest.fn(),
      focusedMonthIndex: 0,
      setFocusedMonthIndex: jest.fn(),
      focusedYearIndex: 0,
      setFocusedYearIndex: jest.fn(),
      yearPositionIndex: 0,
      setYearPositionIndex: jest.fn(),
    };
    return (
      <DatePickerContext.Provider value={contextValue}>
        <CalendarHeader
          displayDate={date}
          onChange={onChangeMonth}
          resetFocusOnHeader={resetFocusOnHeader}
          handleTabPressOnPreviousButton={handleTabPressOnPreviousButton}
          handleTabPressOnHeaderTitle={handleTabPressOnHeaderTitle}
          handleTabPressOnNextButton={handleTabPressOnNextButton}
        />
        ;
      </DatePickerContext.Provider>
    );
  };

  it('should change century years when click previous', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-left')).toBeInTheDocument();
    expect(
      getByText(
        `${displayDate.getFullYear()} - ${displayDate.getFullYear() + 11}`
      )
    ).toBeInTheDocument();

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    const newLowerYear = displayDate.getFullYear() - 12;
    await waitFor(() =>
      expect(
        getByText(`${newLowerYear} - ${newLowerYear + 11}`)
      ).toBeInTheDocument()
    );
    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    const newerLowerYear = newLowerYear - 12;
    await waitFor(() =>
      expect(
        getByText(`${newerLowerYear} - ${newerLowerYear + 11}`)
      ).toBeInTheDocument()
    );

    fireEvent.click(container.querySelector('i.bi-chevron-left')!);
    const newestLowerYear = newerLowerYear - 12;

    await waitFor(() =>
      expect(
        getByText(`${newestLowerYear} - ${newestLowerYear + 11}`)
      ).toBeInTheDocument()
    );
    fireEvent.click(container.querySelector('i.bi-chevron-right')!);
    await waitFor(() =>
      expect(
        getByText(`${newerLowerYear} - ${newerLowerYear + 11}`)
      ).toBeInTheDocument()
    );
  });
  it('should change century years when click next', async () => {
    const { getByText, container } = render(<CalendarHeaderComponent />);

    expect(container.querySelector('i.bi-chevron-right')).toBeInTheDocument();
    expect(
      getByText(
        `${displayDate.getFullYear()} - ${displayDate.getFullYear() + 11}`
      )
    ).toBeInTheDocument();
    const newUpperYear = displayDate.getFullYear() + 12;
    fireEvent.click(container.querySelector('i.bi-chevron-right')!);

    await waitFor(() =>
      expect(
        getByText(`${newUpperYear} - ${newUpperYear + 11}`)
      ).toBeInTheDocument()
    );
  });
});
