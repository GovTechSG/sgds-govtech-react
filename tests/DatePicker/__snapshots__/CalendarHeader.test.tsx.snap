// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader should have default html structure 1`] = `
<DocumentFragment>
  <div
    class="text-center d-flex justify-content-between"
  >
    <button
      aria-label="previous day"
    >
      <i
        class="bi bi-chevron-left"
      />
    </button>
    <button
      aria-live="polite"
      id="id-grid-label"
    >
      March 2022
    </button>
    <button
      aria-label="next day"
    >
      <i
        class="bi bi-chevron-right"
      />
    </button>
  </div>
</DocumentFragment>
`;
